src/python/pants/core/util_rules/adhoc_process_support.py

```diff
@@ -542,6 +552,56 @@ async def run_adhoc_process(
     return AdhocProcessResult(result, adjusted)
 
 
+async def compute_workspace_invalidation_hash(path_globs: PathGlobs) -> str:
+    # Helper to convert a PathMetadataKind to an integer.
+    def kind_to_int(kind: PathMetadataKind) -> int:
+        if kind == PathMetadataKind.FILE:
+            return 0
+        elif kind == PathMetadataKind.DIRECTORY:
+            return 1
+        elif kind == PathMetadataKind.SYMLINK:
+            return 2
+        else:
+            raise TypeError(f"Unable to discriminate instance of `PathMetadataKind`: {repr(kind)}")
+
+    # Compute a stable hash value for a `PathMetadata` since the hash value should be stable when used outside the process.
+    # (for example, in the cache). The `__hash__` dunder method computes an unstable hash
+    # which can and does vary across different process invocations.)
+    #
+    # Ultimately, this may be more of an intellectual correctness point than a necessity since this code will likely never
+    # see the same `m.modified` value. It does matter, however, for a single user to see the same
+    # behavior across process invocations if pantsd restarts.
+    def stable_metadata_bytes(m: PathMetadata | None) -> bytes:
+        if m is None:
+            return b""
+
+        return struct.pack(
+            "sBQBIsss",
+            m.path,
+            kind_to_int(m.kind),
+            m.length,
+            1 if m.is_executable else 0,
+            m.unix_mode if m.unix_mode is not None else 0,
+            m.created.isoformat() if m.created is not None else "",
+            m.modified.isoformat() if m.modified is not None else "",
```

Comment:
Note: I intentionally did not hash `accessed` time here, just `created` and `modified` times.